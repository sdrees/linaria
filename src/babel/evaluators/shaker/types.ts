import { types as t } from '@babel/core';

export type NodeOfType<T> = Extract<t.Node, { type: T }>;

export type NodeType = t.Node['type'] | keyof t.Aliases;

export type VisitorAction = 'ignore' | void;

export type Visitor<TNode extends t.Node> = <TParent extends t.Node>(
  node: TNode,
  parent: TParent | null,
  parentKey: t.VisitorKeys[TParent['type']] | null,
  listIdx: number | null
) => VisitorAction;

export type Visitors = { [TMethod in NodeType]?: Visitor<NodeOfType<TMethod>> };

export type IdentifierHandlerType = 'declare' | 'keep' | 'refer';

export type IdentifierHandlers = {
  [key in IdentifierHandlerType]: [NodeType, ...string[]][];
};
