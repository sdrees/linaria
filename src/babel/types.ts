import { types as t, TransformOptions } from '@babel/core';
import { NodePath } from '@babel/traverse';
import { StyledMeta } from '../types';

export type JSONValue = string | number | boolean | JSONObject | JSONArray;

export interface JSONObject {
  [x: string]: JSONValue;
}

export interface JSONArray extends Array<JSONValue> {}

export type Serializable = JSONArray | JSONObject;

export enum ValueType {
  COMPONENT,
  LAZY,
  FUNCTION,
  VALUE,
}

export type Value = Function | StyledMeta | string | number;

export type ValueCache = Map<t.Expression | string, Value>;

export type ComponentValue = {
  kind: ValueType.COMPONENT;
  ex: NodePath<t.Expression> | t.Expression | string;
};

export type LazyValue = {
  kind: ValueType.LAZY;
  ex: NodePath<t.Expression> | t.Expression | string;
};

export type FunctionValue = {
  kind: ValueType.FUNCTION;
  ex: any;
};

export type EvaluatedValue = {
  kind: ValueType.VALUE;
  value: Value;
};

export type ExpressionValue =
  | ComponentValue
  | LazyValue
  | FunctionValue
  | EvaluatedValue;

export type TemplateExpression = {
  styled?: { component: any };
  path: NodePath<t.TaggedTemplateExpression>;
  expressionValues: ExpressionValue[];
};

export type State = {
  queue: TemplateExpression[];
  rules: {
    [selector: string]: {
      className: string;
      displayName: string;
      cssText: string;
      start: Location | null | undefined;
    };
  };
  replacements: Array<{
    original: {
      start: Location;
      end: Location;
    };
    length: number;
  }>;
  index: number;
  dependencies: string[];
  file: {
    opts: {
      cwd: string;
      root: string;
      filename: string;
    };
    metadata: any;
  };
};

export type Evaluator = (
  filename: string,
  options: StrictOptions,
  text: string,
  only: string[] | null
) => [string, Map<string, string[]> | null];

export type EvalRule = {
  test?: RegExp | ((path: string) => boolean);
  action: Evaluator | 'ignore' | string;
};

type ClassNameFn = (hash: string, title: string) => string;

export type StrictOptions = {
  classNameSlug?: string | ClassNameFn;
  displayName: boolean;
  evaluate: boolean;
  ignore?: RegExp;
  babelOptions: TransformOptions;
  disabled?: boolean;
  rules: EvalRule[];
};

export type Location = {
  line: number;
  column: number;
};

type AllNodes = { [T in t.Node['type']]: Extract<t.Node, { type: T }> };

declare module '@babel/core' {
  namespace types {
    type VisitorKeys = {
      [T in keyof AllNodes]: Extract<
        keyof AllNodes[T],
        {
          [Key in keyof AllNodes[T]]: AllNodes[T][Key] extends
            | t.Node
            | t.Node[]
            | null
            ? Key
            : never;
        }[keyof AllNodes[T]]
      >;
    };

    const VISITOR_KEYS: { [T in keyof VisitorKeys]: VisitorKeys[T][] };
    const ALIAS_KEYS: {
      [T in t.Node['type']]: {
        [K in keyof t.Aliases]: AllNodes[T] extends t.Aliases[K] ? K : never;
      }[keyof t.Aliases][];
    };

    const FLIPPED_ALIAS_KEYS: {
      [T in keyof t.Aliases]: t.Aliases[T]['type'][];
    };

    function shallowEqual(actual: object, expected: object): boolean;
  }
}
