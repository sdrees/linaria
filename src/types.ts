import { PluginOptions } from './babel/utils/loadOptions';

export type Replacement = {
  original: { start: Location; end: Location };
  length: number;
};

export type Location = {
  line: number;
  column: number;
};

export type Rules = {
  [className: string]: {
    cssText: string;
    displayName: string;
    start: Location | null;
  };
};

export type Result = {
  code: string;
  sourceMap: Object | null | undefined;
  cssText?: string;
  cssSourceMapText?: string;
  dependencies?: string[];
  rules?: Rules;
  replacements?: Replacement[];
};

export type LinariaMetadata = {
  rules: Rules;
  replacements: Replacement[];
  dependencies: string[];
};

export type Options = {
  filename: string;
  preprocessor?: Preprocessor;
  outputFilename?: string;
  inputSourceMap?: Object;
  pluginOptions?: Partial<PluginOptions>;
};

export type StyledMeta = {
  __linaria: {
    className: string;
    extends: StyledMeta;
  };
};

export type PreprocessorFn = (selector: string, cssText: string) => string;
export type Preprocessor = 'none' | 'stylis' | PreprocessorFn | void;
