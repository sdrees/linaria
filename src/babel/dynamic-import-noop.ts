import { types } from '@babel/core';
import { NodePath } from '@babel/traverse';
import syntax from '@babel/plugin-syntax-dynamic-import';

export default function dynamic({ types: t }: { types: typeof types }) {
  return {
    inherits: syntax,
    visitor: {
      Import(path: NodePath<types.Import>) {
        const noop = t.arrowFunctionExpression([], t.identifier('undefined'));

        path.parentPath.replaceWith(
          t.objectExpression([
            t.objectProperty(t.identifier('then'), noop),
            t.objectProperty(t.identifier('catch'), noop),
          ])
        );
      },
    },
  };
}
