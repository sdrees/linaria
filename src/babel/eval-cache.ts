import { createHash } from 'crypto';

import { debug } from './utils/logger';

const fileHashes = new Map<string, string>();
const evalCache = new Map<string, any>();

const hash = (text: string) =>
  createHash('sha1')
    .update(text)
    .digest('base64');

let lastText: string = '';
let lastHash: string = hash(lastText);

const memoizedHash: typeof hash = text => {
  if (lastText !== text) {
    lastHash = hash(text);
    lastText = text;
  }

  return lastHash;
};

export const clear = () => {
  fileHashes.clear();
  evalCache.clear();
};

export const has = (filename: string, text: string): boolean => {
  const textHash = memoizedHash(text);
  debug(`EVAL CACHE has ${filename} ${textHash}`);

  return fileHashes.get(filename) === textHash;
};

export const get = (filename: string, text: string): any => {
  const textHash = memoizedHash(text);
  debug(`EVAL CACHE get ${filename} ${textHash}`);

  if (fileHashes.get(filename) !== textHash) {
    return undefined;
  }

  return evalCache.get(filename);
};

export const set = (filename: string, text: string, value: any): void => {
  const textHash = memoizedHash(text);
  debug(`EVAL CACHE set ${filename} ${textHash}`);

  fileHashes.set(filename, textHash);
  evalCache.set(filename, value);
};
