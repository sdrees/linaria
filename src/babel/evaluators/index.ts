// TypeScript Version: 3.2

import Module from '../module';
import { StrictOptions } from '../types';

export default function evaluate(
  code: string,
  t: any,
  filename: string,
  options: StrictOptions
) {
  const m = new Module(filename, options);

  m.dependencies = [];
  m.evaluate(code, ['__linariaPreval']);

  return {
    value: m.exports,
    dependencies: m.dependencies,
  };
}
