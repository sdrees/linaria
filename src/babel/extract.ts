/* eslint-disable no-param-reassign */

import { types } from '@babel/core';
import { NodePath, Scope } from '@babel/traverse';
import { expression, statement } from '@babel/template';
import generator from '@babel/generator';
import evaluate from './evaluators';
import getTemplateProcessor from './evaluators/templateProcessor';
import Module from './module';
import {
  State,
  StrictOptions,
  LazyValue,
  ExpressionValue,
  ValueType,
  ValueCache,
} from './types';
import TaggedTemplateExpression from './visitors/TaggedTemplateExpression';

function isLazyValue(v: ExpressionValue): v is LazyValue {
  return v.kind === ValueType.LAZY;
}

function isNodePath(obj: any): obj is NodePath {
  return obj && obj.node !== undefined;
}

function findFreeName(scope: Scope, name: string): string {
  // By default `name` is used as a name of the function …
  let nextName = name;
  let idx = 0;
  while (scope.hasBinding(nextName, false)) {
    // … but if there is an already defined variable with this name …
    // … we are trying to use a name like wrap_N
    idx += 1;
    nextName = `wrap_${idx}`;
  }

  return nextName;
}

// All exported values will be wrapped with this function
const expressionWrapperTpl = statement(`
  const %%wrapName%% = (fn) => {
    try {
      return fn();
    } catch (e) {
      return e;
    }
  };
`);

const expressionTpl = expression(`%%wrapName%%(() => %%expression%%)`);
const exportsLinariaPrevalTpl = statement(
  `exports.__linariaPreval = %%expressions%%`
);

function addLinariaPreval(
  path: NodePath<types.Program>,
  lazyDeps: Array<types.Expression | string>
): types.Program {
  // Constant __linariaPreval with all dependencies
  const wrapName = findFreeName(path.scope, '_wrap');
  const statements = [
    expressionWrapperTpl({ wrapName }),
    exportsLinariaPrevalTpl({
      expressions: types.arrayExpression(
        lazyDeps.map(expression => expressionTpl({ expression, wrapName }))
      ),
    }),
  ];

  const programNode = path.node;
  return types.program(
    [...programNode.body, ...statements],
    programNode.directives,
    programNode.sourceType,
    programNode.interpreter
  );
}

export default function extract(_babel: any, options: StrictOptions) {
  const process = getTemplateProcessor(options);

  return {
    visitor: {
      Program: {
        enter(path: NodePath<types.Program>, state: State) {
          // Collect all the style rules from the styles we encounter
          state.queue = [];
          state.rules = {};
          state.index = -1;
          state.dependencies = [];
          state.replacements = [];

          // Invalidate cache for module evaluation to get fresh modules
          Module.invalidate();

          // We need our transforms to run before anything else
          // So we traverse here instead of a in a visitor
          path.traverse({
            // Identifier: p => Identifier(p, state, options),
            TaggedTemplateExpression: p =>
              TaggedTemplateExpression(p, state, options),
          });

          const lazyDeps = state.queue.reduce(
            (acc, { expressionValues: values }) => {
              acc.push(
                ...values
                  .filter(isLazyValue)
                  .map(v => (isNodePath(v.ex) ? v.ex.node : v.ex))
              );
              return acc;
            },
            [] as Array<types.Expression | string>
          );

          let lazyValues: any[] = [];

          if (lazyDeps.length > 0) {
            const program = addLinariaPreval(path, lazyDeps);
            const { code } = generator(program);
            const evaluation = evaluate(
              code,
              types,
              state.file.opts.filename,
              options
            );

            state.dependencies.push(...evaluation.dependencies);
            lazyValues = evaluation.value.__linariaPreval || [];
          }

          const valueCache: ValueCache = new Map();
          lazyDeps.forEach((key, idx) => valueCache.set(key, lazyValues[idx]));

          state.queue.forEach(item => process(item, state, valueCache));
        },
        exit(_: any, state: State) {
          if (Object.keys(state.rules).length) {
            // Store the result as the file metadata
            state.file.metadata = {
              linaria: {
                rules: state.rules,
                replacements: state.replacements,
                dependencies: state.dependencies,
              },
            };
          }

          // Invalidate cache for module evaluation when we're done
          Module.invalidate();
        },
      },
    },
  };
}
