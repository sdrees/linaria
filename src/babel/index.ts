import loadOptions, { PluginOptions } from './utils/loadOptions';

export default function linaria(context: any, options: PluginOptions) {
  return options.disabled
    ? {}
    : {
        plugins: [[require('./extract'), loadOptions(options)]],
      };
}
