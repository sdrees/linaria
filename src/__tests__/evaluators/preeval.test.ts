import * as babel from '@babel/core';
import dedent from 'dedent';
import { join } from 'path';
import serializer from '../../__utils__/linaria-snapshot-serializer';
import { StrictOptions } from '../../babel/types';

expect.addSnapshotSerializer(serializer);

const options: Partial<StrictOptions> = {
  displayName: true,
  evaluate: true,
};

const transpile = async (input: string) =>
  (await babel.transformAsync(input, {
    babelrc: false,
    presets: [[require.resolve('../../babel/evaluators/preeval'), options]],
    plugins: [
      '@babel/plugin-proposal-class-properties',
      '@babel/plugin-syntax-jsx',
    ],
    filename: join(__dirname, 'app/index.js'),
  }))!;

it('preserves classNames', async () => {
  const { code } = await transpile(
    dedent`
      import { styled } from 'linaria/react';
      
      const Component = styled.div\`\`;
      `
  );

  expect(code).toMatchSnapshot();
});

it('replaces functional component', async () => {
  const div = '<div>{props.children}</div>';
  const { code } = await transpile(
    dedent`
      import React from 'react';
      
      const Component = (props) => ${div};
      `
  );

  expect(code).toMatchSnapshot();
});

it('replaces class component', async () => {
  const div = '<div>{props.children}</div>';
  const { code } = await transpile(
    dedent`
      import React from 'react';
      
      class Component extends React.PureComponent {
        render() {
          return ${div};
        }
      }
      `
  );

  expect(code).toMatchSnapshot();
});

it('replaces constant', async () => {
  const div = '<div>{props.children}</div>';
  const { code } = await transpile(
    dedent`
      import React from 'react';
      
      const tag = ${div};
      
      const Component = (props) => tag;
      `
  );

  expect(code).toMatchSnapshot();
});
