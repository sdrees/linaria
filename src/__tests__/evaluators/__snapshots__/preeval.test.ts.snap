// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preserves classNames 1`] = `
"import { styled } from 'linaria/react';
const Component =
/*linaria ca1zxek Component Component_ca1zxek*/
styled.div\`\`;"
`;

exports[`replaces class component 1`] = `
"import React from 'react';

function Component() {
  return <></>;
}"
`;

exports[`replaces constant 1`] = `
"import React from 'react';
const tag = <></>;

const Component = props => tag;"
`;

exports[`replaces functional component 1`] = `
"import React from 'react';

const Component = props => {
  return <></>;
};"
`;
